### PR TITLE
[PWX-33946]: [resize-disk]While repl-add operation is in progress do pool resize

### PR DESCRIPTION
**Pull request by ragsingh-px**
**What this PR does / why we need it**:
Automation test for While repl-add operation is in progress do pool resize using resize-disk
**Which issue(s) this PR fixes** (optional)
Closes # 
PWX-33946

**Special notes for your reviewer**:
Testing link:- https://jenkins.pwx.dev.purestorage.com/user/ragsingh@purestorage.com/my-views/view/all/job/Users/job/Raghvendra/job/raghav_pxdv-7-dev/45

### CODE REVIEW

File: tests/basic/pool_function_test.go [hunk 1]
```diff
@@ -833,3 +834,87 @@ var _ = Describe("{PoolExpandAddDiskInMaintenanceMode}", func() {
...
+
+	AfterEach(func() {
+		appsValidateAndDestroy(contexts)
+		EndTorpedoTest()
+	})
+	stepLog := "Expand volume to the pool and pool expansion using resize-disk"
+	It(stepLog, func() {
+		var newRep int64
+		var currRep int64
+		poolToResize = getStoragePool(poolIDToResize)
```
Comment (pp511) [inline tests/basic/pool_function_test.go:866]: You have poolIDToResize in JustBeforeEach. It is not required in the start of this test("It") then.

File: tests/basic/pool_function_test.go [hunk 2]
```diff
@@ -833,3 +834,87 @@ var _ = Describe("{PoolExpandAddDiskInMaintenanceMode}", func() {
...
+		volSelected, err := GetVolumeWithMinimumSize(contexts, 10)
+		log.FailOnError(err, "error identifying volume")
+		opts := volume.Options{
+			ValidateReplicationUpdateTimeout: replicationUpdateTimeout,
+		}
+
+		stepLog = "Expand volume to the intended pool"
```
Comment (alicelyu-px) [inline tests/basic/pool_function_test.go:873]: Please check out example usage of "stepLog"  https://github.com/portworx/torpedo/blob/dda8caa3a1396afd1ec5c369a4a4a3b4cb2ae62c/tests/basic/pool_function_test.go#L189
If you only want to log something out, can simply call `log.InfoD("[the log message]")`.

File: tests/basic/pool_function_test.go [hunk 3]
```diff
@@ -833,3 +834,87 @@ var _ = Describe("{PoolExpandAddDiskInMaintenanceMode}", func() {
...
 	})
 })
+
+var _ = Describe("{PoolVolUpdateResizeDisk}", func() {
+
+	var testrailID = 34542876
+	// testrailID corresponds to: https://portworx.testrail.net/index.php?/tests/view/34542876
+
+	BeforeEach(func() {
+		StartTorpedoTest("PoolVolUpdateResizeDisk", "expand volume to the pool and pool expansion using resize-disk", nil, testrailID)
```
Comment (alicelyu-px) [inline tests/basic/pool_function_test.go:844]: nit: this test description can be improved with clarity.

File: tests/basic/pool_function_test.go [hunk 4]
```diff
@@ -833,3 +834,87 @@ var _ = Describe("{PoolExpandAddDiskInMaintenanceMode}", func() {
...
+	})
+	JustAfterEach(func() {
+		AfterEachTest(contexts)
+	})
+
+	AfterEach(func() {
+		appsValidateAndDestroy(contexts)
+		EndTorpedoTest()
+	})
+	stepLog := "Expand volume to the pool and pool expansion using resize-disk"
```
Comment (alicelyu-px) [inline tests/basic/pool_function_test.go:862]: An `It` block declares a test case, and its first string parameter is the test description, not a "stepLog".
The current usage is not technically "incorrect" for your `stepLog` does explain the purpose of this test (though the wording needs improvement for clarify), but let's aim to use the framework properly.    
Please spend a few minutes reading the documentation https://onsi.github.io/ginkgo/

File: tests/basic/pool_function_test.go [hunk 5]
```diff
@@ -833,3 +834,87 @@ var _ = Describe("{PoolExpandAddDiskInMaintenanceMode}", func() {
...
+		newRep = currRep
+		if currRep == 3 {
+			newRep = currRep - 1
+			err = Inst().V.SetReplicationFactor(volSelected, newRep, nil, nil, true, opts)
+			log.FailOnError(err, fmt.Sprintf("err setting repl factor  to %d for  vol : %s", newRep, volSelected.Name))
+		}
+		log.InfoD(fmt.Sprintf("setting repl factor to %d for vol : %s", newRep+1, volSelected.Name))
+		err = Inst().V.SetReplicationFactor(volSelected, newRep+1, []string{storageNode.Id}, []string{poolToResize.Uuid}, false, opts)
+		log.FailOnError(err, fmt.Sprintf("err setting repl factor  to %d for  vol : %s", newRep+1, volSelected.Name))
+		dash.VerifyFatal(err == nil, true, fmt.Sprintf("vol %s expansion triggered successfully on node %s", volSelected.Name, storageNode.Name))
```
Comment (alicelyu-px) [inline tests/basic/pool_function_test.go:886]: The code above seems to adjust replication factor, why would the dash verify log expect pool expansion triggered?

File: tests/basic/pool_function_test.go [hunk 6]
```diff
@@ -833,3 +834,87 @@ var _ = Describe("{PoolExpandAddDiskInMaintenanceMode}", func() {
...
+		poolToResize = getStoragePool(poolIDToResize)
+		originalSizeInBytes = poolToResize.TotalSize
+		targetSizeInBytes = originalSizeInBytes + 100*units.GiB
+		targetSizeGiB = targetSizeInBytes / units.GiB
+		log.InfoD("Current Size of the pool %s is %d GiB. Trying to expand to %v GiB with type resize-disk", poolIDToResize, poolToResize.TotalSize/units.GiB, targetSizeGiB)
+		triggerPoolExpansion(poolIDToResize, targetSizeGiB, api.SdkStoragePool_RESIZE_TYPE_RESIZE_DISK)
+		resizeErr = waitForOngoingPoolExpansionToComplete(poolIDToResize)
+		dash.VerifyFatal(resizeErr, nil, "Pool expansion does not result in error")
+		//reverting the replication for volume validation
+		if currRep < 3 {
```
Comment (alicelyu-px) [inline tests/basic/pool_function_test.go:914]: Why do you have this  `if currRep < 3` check?

File: tests/basic/pool_function_test.go (hunk 2 above)
Comment (ragsingh-px) [reply tests/basic/pool_function_test.go]: Done

File: tests/basic/pool_function_test.go (hunk 5 above)
Comment (ragsingh-px) [reply tests/basic/pool_function_test.go]: We are verifying the volume expansion triggered, not pool expansion.

File: tests/basic/pool_function_test.go (hunk 6 above)
Comment (ragsingh-px) [reply tests/basic/pool_function_test.go]: removed it. not needed.

File: tests/basic/pool_function_test.go (hunk 3 above)
Comment (ragsingh-px) [reply tests/basic/pool_function_test.go]: changed it.

File: tests/basic/pool_function_test.go [hunk 7]
```diff
@@ -932,3 +933,68 @@ var _ = Describe("{PoolExpandTestLimits}", func() {
...
+		log.FailOnError(err, "error identifying volume")
+		opts := volume.Options{
+			ValidateReplicationUpdateTimeout: replicationUpdateTimeout,
+		}
+		Step("Increase the HA replica of the volume", func() {
+			currRep, err = Inst().V.GetReplicationFactor(volSelected)
+			log.FailOnError(err, fmt.Sprintf("err getting repl factor for  vol : %s", volSelected.Name))
+			newRep = currRep
+			if currRep == 3 {
+				newRep = currRep - 1
```
Comment (alicelyu-px) [inline tests/basic/pool_function_test.go:975]: The `It` description says "Increase the HA replica", however the code is decreasing HA replica if currRep=3, and not making any change if currRep is not exactly 3. Is this intended?

File: tests/basic/pool_function_test.go [hunk 8]
```diff
@@ -932,3 +933,68 @@ var _ = Describe("{PoolExpandTestLimits}", func() {
...
+			targetSizeInBytes = originalSizeInBytes + 100*units.GiB
+			targetSizeGiB = targetSizeInBytes / units.GiB
+			log.InfoD("Current Size of the pool %s is %d GiB. Trying to expand to %v GiB with type resize-disk", poolIDToResize, poolToResize.TotalSize/units.GiB, targetSizeGiB)
+			triggerPoolExpansion(poolIDToResize, targetSizeGiB, api.SdkStoragePool_RESIZE_TYPE_RESIZE_DISK)
+
+			log.InfoD("Wait for expansion to finish")
+			resizeErr := waitForOngoingPoolExpansionToComplete(poolIDToResize)
+			dash.VerifyFatal(resizeErr, nil, "Pool expansion does not result in error")
+			err = ValidateReplFactorUpdate(volSelected, newRep+1)
+			log.FailOnError(err, "error validating repl factor for vol [%s]", volSelected.Name)
```
Comment (alicelyu-px) [inline tests/basic/pool_function_test.go:996]: Have you tested with a setup where rep count is not 3? Wonder what would be the result if rep count is not changed at all from previous step.

**Review by alicelyu-px — CHANGES_REQUESTED** [review]
These test code should be sell-explanatory. However it is a bit confusing that step description says "increase HA" but test code decreases HA. If this is intended, could you please fix the description?

**Comment by ragsingh-px** [comment]
> These test code should be sell-explanatory. However it is a bit confusing that step description says "increase HA" but test code decreases HA. If this is intended, could you please fix the description?

The volume originally has a HA of 2. We are adding a check at first to see if the HA is 3. 
If it is, we decrease it to 2 and then increase it back to 3 again. 
If not, then we are good to increase it by 1.
While this increase is happening, we do the pool resize operation.

File: tests/basic/pool_function_test.go (hunk 7 above)
Comment (ragsingh-px) [reply tests/basic/pool_function_test.go:1383]: Yes

**Comment by alicelyu-px** [comment]
> > These test code should be sell-explanatory. However it is a bit confusing that step description says "increase HA" but test code decreases HA. If this is intended, could you please fix the description?
> 
> The volume originally has a HA of 2. We are adding a check at first to see if the HA is 3. If it is, we decrease it to 2 and then increase it back to 3 again. If not, then we are good to increase it by 1. While this increase is happening, we do the pool resize operation.

Thanks for the explanation! It would be helpful to include these information in a comment section of this text, probably around line 1381.

File: tests/basic/pool_function_test.go [hunk 9]
```diff
@@ -1338,3 +1339,71 @@ var _ = Describe("{CheckPoolLabelsAfterResizeDisk}", func() {
...
+		}
+		Step("Increase the HA replica of the volume", func() {
+			currRep, err = Inst().V.GetReplicationFactor(volSelected)
+			log.FailOnError(err, fmt.Sprintf("err getting repl factor for  vol : %s", volSelected.Name))
+			newRep = currRep
+			// If the HA is 3, reduce it by 1, so that we can increase it later
+			if currRep == 3 {
+				newRep = currRep - 1
+				err = Inst().V.SetReplicationFactor(volSelected, newRep, nil, nil, true, opts)
+				log.FailOnError(err, fmt.Sprintf("err setting repl factor  to %d for  vol : %s", newRep, volSelected.Name))
```
Comment (alicelyu-px) [inline tests/basic/pool_function_test.go:1385]: Do we need to wait and verify the replication factor was changed before proceeding?

File: tests/basic/pool_function_test.go [hunk 10]
```diff
@@ -1338,3 +1339,71 @@ var _ = Describe("{CheckPoolLabelsAfterResizeDisk}", func() {
...
+			targetSizeGiB = targetSizeInBytes / units.GiB
+			log.InfoD("Current Size of the pool %s is %d GiB. Trying to expand to %v GiB with type resize-disk", poolIDToResize, poolToResize.TotalSize/units.GiB, targetSizeGiB)
+			triggerPoolExpansion(poolIDToResize, targetSizeGiB, api.SdkStoragePool_RESIZE_TYPE_RESIZE_DISK)
+
+			log.InfoD("Wait for expansion to finish")
+			resizeErr := waitForOngoingPoolExpansionToComplete(poolIDToResize)
+			dash.VerifyFatal(resizeErr, nil, "Pool expansion does not result in error")
+			err = ValidateReplFactorUpdate(volSelected, newRep+1)
+			log.FailOnError(err, "error validating repl factor for vol [%s]", volSelected.Name)
+
```
Comment (alicelyu-px) [inline tests/basic/pool_function_test.go:1406]: The rep factor change as a result of running this test may not be desired. Shall we reset the rep factor back to the original after test run?

File: tests/basic/pool_function_test.go (hunk 9 above)
Comment (ragsingh-px) [reply tests/basic/pool_function_test.go:1385]: No. The test states to do pool expansion while the replica increase is going on.

Comment (ragsingh-px) [reply tests/basic/pool_function_test.go:1390]: fixed.

File: tests/basic/pool_function_test.go (hunk 10 above)
Comment (ragsingh-px) [reply tests/basic/pool_function_test.go:1416]: Done

**Review by alicelyu-px — APPROVED** [review]

**Pull request merged**